### ISSUE TITLE
Thumbnail offset hardcoded weirdly?

### ISSUE THREAD

**Issue by suvis1** [motioneye-project/motioneye#2766]
Hi!

Ive been having issues that Motioneye generates thumbnails that are not representative of the actual detection event. I checked the code and found out theres a a minimum offset, and also a weirdly doubled offset for precapture?

For example if i have a camera with framerate of 10, and precapture of 40 frames before a detection, we end up with the thumbnail being generated 8 seconds from the video, eventhough the detection is after the 40 precapture frames, meaning the thumbnails 4 seconds late.

I managed to get alot better thumbnails when i changed the offs value to just be offs +=1 so the thumbnails are taken 1 second after detection.

This seems to be done deliberately so maybe im missing the point of this? is there any reason for it to do this? 

https://github.com/motioneye-project/motioneye/blob/03882a18994fc35c22ae169dd1569abb923c7cef/motioneye/mediafiles.py#L349-L363

**Comment by MichaIng** [comment]
Many thanks for your report. Good find, I have no idea where this 4 seconds minimum + doubled offset is coming from. It was last touched here, before which it was 4 seconds maximum offset: https://github.com/motioneye-project/motioneye/commit/b14bbb5

@ccrisan do you know what this is for? Generally I'd just remove any min/max and the `* 2` to just use the user setting as is.

**Comment by suvis1** [comment]
I feel like some delay is still warranted, since for example a thumbnail taken from the moment of detection only shows a very small part of a detection like this:

![image](https://github.com/motioneye-project/motioneye/assets/51976044/0932a7d5-d80a-4793-9532-c4e83faa979d)
I blurred it for privacy, but you can still see that only the head of the person is in the thumbnail on the driveway, adding 1 second more generally gives a lot better results!

![image](https://github.com/motioneye-project/motioneye/assets/51976044/616be5af-4f82-4f75-879f-c6fdb650b9d0)

**Comment by Semmu** [comment]
I have recently also dig into this code and IMHO the best solution would be to create the thumbnail from the exact middle of the video, I think that would be more informative than the beginning of the motion detection/event.

**Comment by zagrim** [comment]
Coming to this from the above linked PR created by @JamBalaya56562:
Like @suvis1, I also think that using the 1st frame of motion event is not generally a good solution, exactly for those reasons shown with the pair of images. But I'm also not at all convinced that the middle of the event would be any better, either. It all depends a lot of the setup and environment.
- In a setup where the events are expected to be short (or max movie length is short) and/or linear, like someone walking across the scene, the middle might well be a good approximation. But how about when there is another person who also walks across the scene so that the event length is roughly doubled (given max length allows that)? The middle would be between those people passing by, possibly showing nothing of interest.
- Any constant number of frames is not good given that the framerate could be 2 or it could be 30.
- Any constant number of seconds into the movie is also likely to miss the "best" frame by a lot, since the best part might be 1 second from the start, or 10 seconds from the start.
- In less controlled environments (like outdoors) the event might start "too early" due to several things being going on around the scene or around it (like a gush of wind causing movement in the scene, or a bird flying away from it just before the actual interesting part), or it might be difficult to set the motion detection parameters and thus the really best frame might be at the start or at the end of the footage.

Too bad that there is likely no way of knowing the frame where the largest amount of "motion" occurs (without performing frame-by-frame difference calculation), even though if the user has set still image capture mode to "Motion triggered (one picture)" Motion takes that one picture exactly from the frame which has the largest amount of motion (that setting being translated to `picture_output=best`). If the user had enabled saving of still images using that mode (when a given movie was saved), the thumbnail could well be the picture saved by Motion, but there's no generic feasible way of knowing which saved picture goes with which movie and with which mode the picture had been saved, and there just might not be any saved images.

So, I guess there's no one good solution for the issue. We might need to have 2 or 3 user selectable modes for generating the thumbnail, and even then the results might be a hit or miss for some users. For example:
1) beginning of motion event (`offset = pre_capture / framerate`),
2) middle of motion event (`offset = get_movie_duration_seconds(full_path) / 2`), 
3) maybe yet some other option?

**Comment by MichaIng** [comment]
I'd just make the number of frames (based on current framerate) or seconds freely selectable, if changing something about it at all.

v0.44.0 btw fixed failing thumbnail generation for movies shorter than 2x pre-capture at least, falling back to 1x pre-capture, and duration/2, if the movie is shorter, which IMO is a good default behavior until we make it selectable.

**Comment by zagrim** [comment]
The middle point is IMO good option to always offer to the users because it is something that cannot otherwise be specified, but indeed it might be good to just have a user selectable number of seconds from the motion event start as the other option.

I'd go specifically with seconds as the measurement unit here, since there has been some confusion and grievance among some users over time on why some of the motion detection settings work with number of frames instead of seconds (like pre-capture). Most of the time the users need to calculate back and forth between frames and seconds in their heads when they figure out their settings. There might also be files with varying framerates, if the camera config has been changed, or if the user has moved other files into the directory.

**Comment by MichaIng** [comment]
Ageed, seconds is probably easier to understand/handle and use consequently across all settings. The only argument against it is the fact that, with default framerate, an offset of 1 second (which was suggested by someone as better default around this discussion) contains the same frame as an offset of 0 seconds. So technically, it makes more sense to go with frames, which is also why pre/post-capture settings in motion (and motionEye) are like this.

EDIT: Actually, the default framerate is 2, meaning 2 frames per second, which is also the lowest possible value. There is also the `minimum_frame_time` setting to reduce the rate of snapshots and for recordings, while it remains at `framerate` for motion-detection. But we do not seem to set this (?).
So I wonder why pre-capture of 1 does result in 2*2=4 seconds thumbnail offset, instead of 2*0.5=1 second 🤔.

**Comment by zagrim** [comment]
I didn't quite follow @MichaIng how the default framerate of ME would in any way affect the choice between seconds and frames related to thumbnail generation. I don't think the _actual_ framerate is 2 fps in most cases (except perhaps the low-end RPi setups when the camera does not allow having separate high-res stream that is saved using `movie_passthrough`), and it is the actual rate that matters, right?
I'd guess Motion uses frames as the unit in pre/post-capture just because that is the most obvious and easy solution for the technical implementation (which is dealing directly with frames anyway). MotionEye has probably originally adopted the same unit to make things simple and avoid having to calculate between seconds and frames (and thus avoid having to round the result of that calculation, where the rounding error might cause anomalies is some special cases (with small values)).

Supporting `minimum_frame_time` might complicate things even more, but as ME does not really support using that option, let's not take it into account now. It is disabled by default, so unless the user sets that via Extra Motion Options, Motion should anyway be using `framerate` setting.

**Comment by MichaIng** [comment]
I think I just misinterpreted things a bit when discussing/reviewing https://github.com/motioneye-project/motioneye/pull/3304/changes:
- I thought 2x pre-capture would be 4 seconds anyway, but it is probably 1 second by default with the default framerate of 2 (frames per second): https://github.com/motioneye-project/motioneye/blob/dev/motioneye/config.py#L2399
- So the reason for the short video failures was not the 2x pre-capture, but that `max(4, offs * 2)` chose always 4 seconds offset, unless pre-capture was raised to more than 4 frames (2 seconds).

So the PR did indeed change the behavior, taking the thumbnail at 1 second offset (0.5 seconds from motion start), instead of 4 seconds (3.5 seconds from motion start) by default. To restore the previous behavior, we'd need to add `4` again to the new `max()` statement. Now, however, it becomes effective only for videos of at least 4 seconds length.

**Comment by zagrim** [comment]
> To restore the previous behavior, we'd need to add 4 again to the new max() statement. Now, however, it becomes effective only for videos of at least 4 seconds length.

But why would we want to do that? This issue has been created and commented on originally only before 0.44.0.
The case @suvis1 described is one where the motion seems to span only a short while. From the pictures it seems that the event length might be roughly 4 s, and with the 4 s pre-capture that was mentioned (40 f/10 fps), the 4 s offset would be applied even with the current video length condition. And that exactly was the issue that was reported: "we end up with the thumbnail being generated 8 seconds from the video"

As a side-note, pre-capture of 40 frames is way over what [Motion docs](https://motion-project.github.io/motion_config.html#pre_capture) recommend (which is 0 to 5 frames), but if the resource requirements of a large pre-capture buffer doesn't cause performance issues, there might be even longer pre-capture buffers (both in frames and in seconds) out there 🤷 

But anyway, in a case where pre-capture is long and the motion event itself is short (and only short or zero post-capture), even the middle point of the video might be too early, even though it is a sane fallback for the case where the calculated offset exceeds the total video length. I didn't think of that earlier, but as long as the offset is calculated using pre-capture length as the base, it shouldn't really matter.  Given the further though-work, I'll adjust my earlier proposal for the options that could be given to the user to allow more control on the thumbnail generation (edit: and have edited this multiple times since based on the discussion below):
1. Configurable absolute offset from motion start: `offset = pre_capture / framerate + N`, where N is user-configurable number of seconds (decimal).
2. Configurable relative offset from motion start: `offset = pre_capture / framerate + (get_movie_duration_seconds(full_path) - (pre_capture + post_capture) / framerate) * P`, where P is user configurable percentage of offset into the motion event.

Option 2 is already pretty elaborate, I know, but I thought that since the duration of the events in many times can vary even a lot and cannot be really anticipated, perhaps a relative offset could then work better than absolute one. At least it would if the motion is linear, has constant speed and only the speed of the moving object varies between events. However, for the users the number of seconds might likely be the easiest to understand.

If motion is non-linear, the speed of the moving object is varying a lot during the event, or there are multiple (logically separate) events in the same file (because they occurred immediately one after another), all calculation based on motion detection parameters and video length would, of course, be in vain. The only good option for those cases would be to use the frame with largest amount of motion for the thumbnail, but that is not really feasible option for bulk processing.

**Comment by MichaIng** [comment]
We could also take the post-capture into account for the relative offset. And wasn't there another setting about a number of frames without motion before the motion event is formally stopped? Maybe this allows to narrow down the window of frames where really something is moving.

You are right, the v0.44.0 change did basically sort of fix or reduce the issue of the OP, but unintentionally, and instead potentially added an issue for suvis1 that didn't exist before. So the question is whether we want to restore the previous behavior that worked well for most for several years, as a quick status quo, or implement a proper option directly.

**Comment by zagrim** [comment]
> And wasn't there another setting about a number of frames without motion before the motion event is formally stopped? Maybe this allows to narrow down the window of frames where really something is moving.

Umm.. yes, there's `event_gap` (https://motion-project.github.io/motion_config.html#event_gap) (in UI "Motion Gap") which is the number of seconds motion is no longer detected and post_capture has run out, after which the motion event is ended. So basically, the video file consists of pre-capture + actual motion + post-capture + event gap, and we'd like to have the thumbnail be taken just from the actual motion. I'll amend by proposal above to also subtract `event_gap` from the video length in option 2. I already had `post_capture` taken into account there.

**Comment by zagrim** [comment]
> You are right, the v0.44.0 change did basically sort of fix or reduce the issue of the OP, but unintentionally, and instead potentially added an issue for suvis1 that didn't exist before. So the question is whether we want to restore the previous behavior that worked well for most for several years, as a quick status quo, or implement a proper option directly.

This issue was created by suvis1 over 3 years old now, so it is exactly about the "previour behavior" 😃

**Comment by MichaIng** [comment]
Err right, also it suggests 1 second offset, while 4 seconds minus pre-capture is 3.5 seconds by default, so indeed quite long after motion start.

So we probably better skip adding the 4s back in, unless someone is complaining about the new default 1s (0.5s after motion start) being an issue, and instead inplement a setting directly, when someone finds time.

We might want to verift that the event gap is really part of the recording, and not cut away at the end motion-internally. Main purpose is most likely to not have multiple motion events and individual recordings, if the person/bird/whatever stands still for a single frame, as kind of a "is this event really over or just paused?" check. That would also work if it overlaped with post-capture, and everything after post-capture was cut from the recording, it the gap passed without new motion.

**Comment by JamBalaya56562** [comment]
@MichaIng re: verifying whether the event gap is part of the recording — **it is not**. The file merely stays open across it. Checked in motion 4.5.1's source and measured on a live setup.

### The direct test

Everything identical, only `event_gap` differs:

| | `event_gap` | movie duration | last recorded frame | file closed | Δ |
|---|---|---|---|---|---|
| **A** | 5 s | 21.85 s | `10:02:31` `Postcap` | `10:02:36` | **5 s** |
| **B** | 20 s | 24.83 s | `10:08:35` `Postcap` | `10:08:55` | **20 s** |

Raising `event_gap` by 15 s did **not** add 15 s of video. The gap elapses *after* the last recorded frame, and only then does the file close.

At `log_level debug` motion burns the frame type into every frame it saves, so the recordings document themselves — first frame on top, last frame below:

| A — `event_gap 5` | B — `event_gap 20` |
|---|---|
| <img src="https://raw.githubusercontent.com/JamBalaya56562/motioneye/issue-2766-assets/issue-2766-A-labels.png" width="330"> | <img src="https://raw.githubusercontent.com/JamBalaya56562/motioneye/issue-2766-assets/issue-2766-B-labels.png" width="330"> |
| recorded `10:02:09` → `10:02:31`, then file closed `10:02:36` | recorded `10:08:11` → `10:08:35`, then file closed `10:08:55` |

### Your "is this event really over or just paused?" case

A 10 s pause inside one event, `event_gap 20`. Result: one movie, and the pause is **elided** — two *adjacent* frames jump 8 s of wall clock:

| Adjacent frames at the splice | Whole recorded timeline (1 frame per playback second) |
|---|---|
| <img src="https://raw.githubusercontent.com/JamBalaya56562/motioneye/issue-2766-assets/issue-2766-C-splice.png" width="230"> | <img src="https://raw.githubusercontent.com/JamBalaya56562/motioneye/issue-2766-assets/issue-2766-C-timeline.png" width="230"> |

Control: same run with `event_gap 5` (shorter than the 10 s pause) gives **two** movies, each closed 5 s after its own last frame — so the parameter was genuinely in effect.

### What this means for the formula

Subtracting `event_gap` over-subtracts, because that time was never in the movie. At the 30 s default it goes negative for any movie shorter than ~30 s — most of them.

The goal (skip the pre-roll so the subject is actually in frame) is unaffected; it just wants an offset derived from the recorded content and clamped, rather than from `event_gap`.

<details>
<summary><b>Source references</b> — motion 4.5.1 (<code>09afbfdd</code>)</summary>

- Once post-capture is exhausted, frames get `IMAGE_PRECAP` **without** `IMAGE_SAVE` — [`motion.c:2424-2426`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/motion.c#L2424-L2426)
- `process_image_ring()` stops at the first frame not flagged `IMAGE_SAVE` — [`motion.c:650-652`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/motion.c#L650-L652)

So gap frames are never handed to the encoder; they aren't "cut away at the end", they're never added.

Your guess at the purpose is right, and motion goes further: when motion resumes after a lull it rewinds the movie clock **so the pause leaves no hole** — [`motion.c:2389-2395`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/motion.c#L2389-L2395), comment: *"we reset the start movie time so that we do not get a pause in the movie"*.

The per-frame type labels shown above come from [`motion.c:657-683`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/motion.c#L657-L683) — drawn inside `process_image_ring()`, i.e. only on frames that actually get encoded.
</details>

<details>
<summary><b>Test setup and full matrix</b></summary>

Stock `dev`, motion 4.5.1, ffmpeg 5.1.9, x86 container. JPEG netcam served locally, moving on a fixed schedule so the motion event is reproducible. `framerate 10`, `pre_capture 10`, `post_capture 10`, `minimum_motion_frames 5`, `noise_tune off`, `movie_codec mp4`, `movie_max_time 0`, `log_level debug`.

| Test | `event_gap` | pause | movies | duration | last frame | file closed | Δ |
|---|---|---|---|---|---|---|---|
| A | 5 s | – | 1 | 21.85 s | `10:02:31` | `10:02:36` | 5 s |
| B | 20 s | – | 1 | 24.83 s | `10:08:35` | `10:08:55` | 20 s |
| C | 20 s | 10 s | **1** | 21.35 s | – | – | pause elided |
| D | 5 s | 10 s | **2** | 14.14 / 6.98 s | `10:16:18` | `10:16:23` | 5 s |

`framerate 10` was chosen deliberately to stay clear of the low-fps container rewrite noted below.
</details>

<details>
<summary><b>Two other things affecting the <code>pre_capture / framerate</code> term</b></summary>

1. **The pre-roll is `pre_capture + minimum_motion_frames` frames, not `pre_capture`.** The ring is sized [`pre_capture + minimum_motion_frames`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/motion.c#L1722) and the *whole* ring is marked `IMAGE_SAVE` on trigger ([`motion.c:2405`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/motion.c#L2405)). Visible in test C: the frames after the splice are labelled `Precap`.
2. **The movie's fps is the *measured* rate, not the configured one** — `movie_fps = lastrate`, floored at 2 ([`event.c:912-918`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/event.c#L912-L918)). A camera that can't keep up makes `pre_capture / framerate` an over-estimate of the pre-roll in seconds.

Also worth knowing: for mp4/mov/flv/mpeg4/hevc at fps ≤ 5 — which includes motionEye's default `framerate 2` — motion rewrites the container fps to 10 ([`ffmpeg.c:713-721`](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/ffmpeg.c#L713-L721)).
</details>

<details>
<summary><b>Caveats</b></summary>

- **motion 4.5.1 only.** 4.6+ rewrote this code (`movie.cpp`); it should be re-checked there before anything ships if motionEye supports those versions.
- I measured a **JPEG netcam** in a container. The `IMAGE_SAVE` gate sits above the capture backend, so I'd expect this to be camera-independent, but that part is inference, not measurement.
- Durations carry ±1 frame of ffprobe slop, and motionEye truncates duration to `int` ([`mediafiles.py:445`](https://github.com/motioneye-project/motioneye/blob/dev/motioneye/mediafiles.py#L445)) — a sub-second movie reports 0, which any offset formula needs to survive.
</details>

Happy to implement the setting once you've settled on its shape.

**Comment by MichaIng** [comment]
Thank you very much for the quick check. So either hardcoded offset from pre-capture on, or relative offset between pre- and post-capture, using half movie length as fallback like now, if the movie is too short (copied, or pre/post-capture settings changed).

The relative offset would work reliably across all settings ranges, but might also result in an unexpectedly long offset in case of long motion events. Personally, for most common use cases, I'd use a fixed ~2 seconds offset, regardless how long the movie is. But that does not work well when capturing a bird flying by is also wanted. Offering two dedicated settings is probably overkill? So in that case, I'd go with the relative setting, which just assures best to catch any motion in every case.

But you might want to give us some time to finish your open PRs first. Due to vacation I was not able to properly review and especially test them, yet.

**Comment by zagrim** [comment]
👍 Yeah, thank you @JamBalaya56562 for checking that, it seems that my interpretation of the Motion docs was not good enough then. They state that "An event ends right after no more motion is detected and post_capture is over." and also "Once the movement has stopped being detected, the post_capture starts. Then once that is finished, the timer starts as specified in event_gap. If any motion is detected before the timer finishes, the existing event continues and the timer is reset. If no motion is detected during the time, the event ends." and from that I drew my incorrect conclusion, but the implementation is clearly more clever than I thought.
I can also confirm your results from my own saved recordings, which do not all have a 5 sec non-motion at the end although I have `event_gap` set at 5. I'll fix my proposal once again.

**Comment by zagrim** [comment]
This:
> Also worth knowing: for mp4/mov/flv/mpeg4/hevc at fps ≤ 5 — which includes motionEye's default framerate 2 — motion rewrites the container fps to 10 ([ffmpeg.c:713-721](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/ffmpeg.c#L713-L721)).

That's an interesting detail and I was able to confirm also that in practise. However, because thumbnails are generated using time-based offset, it shouldn't (luckily) matter what the saved video file framerate is (as long as there's no jumping and playback speed is not altered). That also allows thumbnail generation to work just as fine when the stream used for motion detection is not the same that gets saved (using `movie_passthrough on ).

This, however:
> The movie's fps is the measured rate, not the configured one — movie_fps = lastrate, floored at 2 ([event.c:912-918](https://github.com/Motion-Project/motion/blob/09afbfdd6ab5cde58286e93561b862cdb9509820/src/event.c#L912-L918)). A camera that can't keep up makes pre_capture / framerate an over-estimate of the pre-roll in seconds.

That might indeed affect the accuracy of thumbnail frame selection, but I guess this goes to the category of things we just can't take into account, because that would require first querying the file metadata using `ffprobe` or something like that, causing quite significant overhead per file.

**Comment by zagrim** [comment]
> Offering two dedicated settings is probably overkill? So in that case, I'd go with the relative setting, which just assures best to catch any motion in every case.

Adding a configuration parameter for the user to select a suitable thumbnail selection mode would of course mean somewhat more changes due to changes to UI and config management etc, but my expectation is that there's no single "good enough" way to pick the thumbnail frame for all users and all their varying use cases. But I agree that the relative option should at least work more reliably when the sources and types of motion are varying.

**Comment by Semmu** [comment]
Hi, sorry to interrupt this very technical discussion here but I think it's worth thinking about the end-user UX perspective as well.

MotionEye is a very powerful tool which could be scary for less-technical people at first, so I think it would be nice to default to the middle-of-video thumbnailing (just to honor Occam's Razor, as that is the simplest solution which should work in most setups) and maybe additionally present the advanced settings to people who want to tweak their setup and parameters more in-depth.

What do you think?

**Comment by zagrim** [comment]
Well, if we implemented what above was called "the relative setting" (what I dubbed "Configurable relative offset from motion start" in an earlier comment) and set a default value of 50% for it, that would indeed be what you @Semmu are asking for. How the value for that setting should be displayed in the UI is of course a question of its own for it to be as intuitive as possible for the users.

The hard question is anyway the default value, which optimally should suite the most of the use cases. FWIW, when I think my "production setup", the middle point would be good in many cases, so I'd be ok with that.
